### PR TITLE
sending faction statistics using POST requests at the end of the game

### PR DESCRIPTION
**Pull request by nikodemak**
adds ability to send game statistics to a 3rd party server through a POST request configurable using a registry key (on the server side)

### CODE REVIEW

File: VS2017/Server.vcxproj [hunk 1]
```diff
@@ -77,7 +77,7 @@
     </Midl>
     <ClCompile>
       <Optimization>Disabled</Optimization>
-      <AdditionalIncludeDirectories>..\src\lib\AllegianceSecurity;..\src\lib\Steam;.\;..\src\guids;..\src\zlib;..\src\_Utility;..\src\Igc;..\src\Inc;..\src\AGC;..\src\test\Inc;..\src\lobby;..\src\Zone;.\$(IntDir)..\Igc;.\$(IntDir)..\FedSrv;.\$(IntDir)..\AGC;.\$(IntDir)..\Guids;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>..\src\Lib\nlohmann\json\include;..\src\lib\AllegianceSecurity;..\src\lib\Steam;.\;..\src\guids;..\src\zlib;..\src\_Utility;..\src\Igc;..\src\Inc;..\src\AGC;..\src\test\Inc;..\src\lobby;..\src\Zone;.\$(IntDir)..\Igc;.\$(IntDir)..\FedSrv;.\$(IntDir)..\AGC;.\$(IntDir)..\Guids;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
Comment (BackTrak) [inline VS2017/Server.vcxproj:80]: Looks like you may have added some local paths here by mistake?

File: src/FedSrv/fsmission.cpp [hunk 2]
```diff
@@ -10,6 +10,8 @@
 
 #include "pch.h"
 
+using json = nlohmann::json;
```
Comment (BackTrak) [inline src/FedSrv/fsmission.cpp:13]: Same here, I don't think this will compile on someone else's machine.

File: src/FedSrv/fsmission.h [hunk 3]
```diff
@@ -11,6 +11,10 @@
 #ifndef _FSMISSION_H_
 #define _FSMISSION_H_
 #include "pch.h"
+#include "json.hpp"
+#include "../Lib/cpp-httplib/httplib.h"
```
Comment (BackTrak) [inline src/FedSrv/fsmission.h]: Any reason to use this instead of the existing MaClient (mbedthis HTTP client)?

Comment (nikodemak) [reply src/FedSrv/fsmission.h]: mbedthis doesn't allow changing the encoding header and by default it uses urlencoded while we want json
additionally the source code for it doesn't seem to be available anymore and some of the functions in client.h are not included in the .lib file for it
i actually want to replace mbedthis everywhere else too to not have 2 libraries doing the same thing

Comment (BackTrak) [reply src/FedSrv/fsmission.h]: Got it, thank you! Will you be able to complete the replacement as part of this work?

Comment (nikodemak) [reply src/FedSrv/fsmission.h]: yeah i've created a separate pull request for it which only needs adding debug logging and testing if everything works. i can merge it into this one if that would be better.

Comment (kgersen) [reply src/FedSrv/fsmission.h]: I've always wondered why Windows [wininet](https://learn.microsoft.com/en-us/windows/win32/wininet/http-sessions)  wasn't used for MaClient ?

Comment (nikodemak) [reply src/FedSrv/fsmission.h]: [this page](https://learn.microsoft.com/en-us/windows/win32/wininet/wininet-vs-winhttp) says that wininet cannot be used inside a service and fedsrv as well as lobby can be set up as a service. other that that i have no idea though, you could ask imago but he might not remember either

**Comment by nikodemak** [comment]
​

**Comment by nikodemak** [comment]
new year, new PR merge?

File: src/FedSrv/FedSrv.CPP
Comment (BackTrak) [inline src/FedSrv/FedSrv.CPP:1]: Does this work if the registry setting is missing, or will it crash?

File: src/FedSrv/FedSrv.CPP [hunk 4]
```diff
@@ -9584,6 +9584,11 @@ void LoadSettings()
 		DWORD cbLeaderboardUpdateUrl = sizeof(g.szLeaderboardUpdateUrl);
 		RegQueryValueExA(hKey, "LeaderboardUpdateUrl", NULL, &dwType, (unsigned char*)g.szLeaderboardUpdateUrl, &cbLeaderboardUpdateUrl);
 
+        // Loran - Game statistics logging
+        g.szGameStatsUpdateUrl[0] = '\0';
+        DWORD cbGameStatsUpdateUrl = sizeof(g.szGameStatsUpdateUrl);
+        RegQueryValueExA(hKey, "GameStatsUpdateUrl", NULL, &dwType, (unsigned char*)g.szGameStatsUpdateUrl, &cbGameStatsUpdateUrl);
```
Comment (BackTrak) [inline src/FedSrv/FedSrv.CPP:9590]: Does this work if the registry setting is missing, or will it crash?

File: src/FedSrv/fsmission.cpp [hunk 5]
```diff
@@ -2730,6 +2732,18 @@ void CFSMission::MakeOverrideTechBits()
     }
 }
 
+// Loran - make a POST request with JSON data to export game data
+void POSTGameStats(std::string& url, std::string& postData)
+{
+    std::string host = url.substr(0, url.find_first_of("/", 8));
```
Comment (BackTrak) [inline src/FedSrv/fsmission.cpp]: Does this work if the registry setting is missing, or will it crash?

File: src/FedSrv/FedSrv.CPP (hunk 4 above)
Comment (nikodemak) [reply src/FedSrv/FedSrv.CPP:9590]: good catch, it did but latest commit should fix it.

**Review by BackTrak — DISMISSED** [review]
Hi Loran, could you please add some guards around your code so that if the registry key isn't set, then the new code will be bypassed?

File: src/FedSrv/fsmission.cpp (hunk 5 above)
Comment (nikodemak) [reply src/FedSrv/fsmission.cpp]: already done

**Review by BackTrak — APPROVED** [review]

**Pull request merged**